rasa/utils/tensorflow/callback.py

```diff
@@ -0,0 +1,66 @@
+from pathlib import Path
+from typing import Dict, Text, Any, Optional
+
+import logging
+import tensorflow as tf
+from tqdm import tqdm
+
+import rasa.shared.utils.io
+
+logger = logging.getLogger(__name__)
+
+
+class RasaTrainingLogger(tf.keras.callbacks.Callback):
+    def __init__(self, epochs: int, silent: bool) -> None:
+        super().__init__()
+
+        disable = silent or rasa.shared.utils.io.is_logging_disabled()
+        self.progress_bar = tqdm(range(epochs), desc="Epochs", disable=disable)
+
+    def on_epoch_end(self, epoch: int, logs: Optional[Dict[Text, Any]] = None) -> None:
+        self.progress_bar.update(1)
+        self.progress_bar.set_postfix(logs)
+
+
+class RasaModelCheckpoint(tf.keras.callbacks.Callback):
+    def __init__(self, checkpoint_dir: Path) -> None:
+        super().__init__()
+
+        self.checkpoint_file = checkpoint_dir / "checkpoint.tf_model"
+        self.best_metrics_so_far: Optional[Dict[Text, Any]] = {}
+
+    def on_epoch_end(self, epoch: int, logs: Optional[Dict[Text, Any]] = None) -> None:
+        if self._does_model_improve(logs):
+            logger.debug(f"Creating model checkpoint at epoch={epoch + 1} ...")
+            self.model.save_weights(
+                self.checkpoint_file, overwrite=True, save_format="tf"
+            )
+
+    def _does_model_improve(self, current_results: Dict[Text, Any]) -> bool:
+        # Initialize best_metrics_so_far with the first results
+        if not self.best_metrics_so_far:
+            keys = filter(
+                lambda k: True
+                if (k.startswith("val") and (k.endswith("_acc") or k.endswith("_f1")))
+                else False,
+                current_results.keys(),
+            )
```

Comment:
would it be simpler to create 2 dict inside the model? One for simply outputting metrics and one for validation purposes?

Reply:
We actually have an issue to update our metrics. I think Thomas is working on it. Would do this change along the metric update and not inside this PR.